### PR TITLE
Basic diplomacy

### PR DESCRIPTION
**Pull request by Dilvish-fo**
This moves AI diplomacy into a dedicated module and slightly advances the actual diplomatic dealings.  The AI will now keep a log of peace requests and war declarations.  It now has an 'attitude' regarding the respective other empires, largely shaped by its aggression level but also affected by the diplomatic history between the empires.   Its response to a peace request will be controlled by this attitude.  The AI will also now have a more varied set of responses to peace requests, drawn from the stringtables.

There is obviously much more that can and will be done going forward.  I've tested this moderately & it seems to work fine, but I thought I'd post it for review & feedback before committing.  (Plus, it involves a change in the executables and so I thought I'd wait until closer to Vezzra's weekly build to commit.)

### CODE REVIEW

**Comment by Cjkjvfnby** [comment]
Can you wait for a week before merging it?  I have some ideas how to organize code in other way, but have no time now.  (May be merge c++ part now.)

**Comment by Dilvish-fo** [comment]
So is it just the creation of the DiplomaticCorp module that you'd like me to hold off on?

**Comment by Vezzra** [comment]
> I thought I'd wait until closer to Vezzra's weekly build to commit.

This Mondays test builds will most likely be produced earlier as usual, see the [information I posted here](http://www.freeorion.org/forum/viewtopic.php?f=2&t=9443).

File: default/AI/AIstate.py [hunk 1]
```diff
@@ -1091,3 +1093,20 @@ def split_new_fleets(self): # pylint: disable=invalid-name
...
+        peace_requests.setdefault(log_index, []).append(fo.currentTurn())
+
+    def log_war_declarations(self, initiating_empire_id, recipient_empire_id):
+        """Keep a record of war declarations made or received by this empire."""
+
+        # if war declaration is made on turn 1, don't hold it against them
+        if fo.currentTurn() == 1:
+            return
+        war_declarations = self.diplomatic_logs.setdefault('war_declarations', {})
+        log_index = (initiating_empire_id, recipient_empire_id)
```
Comment (Cjkjvfnby) [inline default/AI/AIstate.py:1111]: Is `recipient_empire_id` matters?

**Comment by Cjkjvfnby** [comment]
I want to store history in more simple way. maybe list of 4 items tuple (type, turn, initiating_empire_id, recipient_empire_id or class instance of diplomatic corps.

File: default/AI/DiplomaticCorp.py [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+import random
+
+import freeOrionAIInterface as fo  # pylint: disable=import-error
+import FreeOrionAI as foAI
+import AIstate
+from freeorion_tools import UserString, UserStringList, chat_on_error, print_error
+
+
+@chat_on_error
+def handle_diplomatic_message(message):  # pylint: disable=invalid-name
```
Comment (Cjkjvfnby) [inline default/AI/DiplomaticCorp.py:12]: why this name is invalid?

File: default/AI/DiplomaticCorp.py [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+    """Handle a diplomatic message update from the server,
+    such as if another player declares war, accepts peace, or cancels a proposed peace treaty.
+    :param message:
+    :return:
+    """
+    print "Received diplomatic %s message from empire %s to empire %s" % (message.type, message.sender, message.recipient)
+    print "my empire id: %s" % fo.empireID()
+    if message.recipient != fo.empireID():
```
Comment (Cjkjvfnby) [inline default/AI/DiplomaticCorp.py:20]: Empire has access to all diplomatic messages in whole universe?

File: default/AI/DiplomaticCorp.py [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+              (len(possible_acknowledgments), acknowledgement)
+        fo.sendChatMessage(proposal_sender_player, acknowledgement)
+        attitude = evaluate_diplomatic_attitude(message.sender)
+        if attitude > 0:
+            reply_text = random.choice(UserStringList("AI_PEACE_PROPOSAL_RESPONSES_YES_" + suffix + "_LIST"))
+            diplo_reply = fo.diplomaticMessage(reply_sender, reply_recipient, fo.diplomaticMessageType.acceptProposal)
+        else:
+            reply_text = random.choice(UserStringList("AI_PEACE_PROPOSAL_RESPONSES_NO_" + suffix + "_LIST"))
+            diplo_reply = None
+        fo.sendChatMessage(proposal_sender_player, reply_text)
```
Comment (Cjkjvfnby) [inline default/AI/DiplomaticCorp.py:40]: IMHO Peace request that was accepted should not increase irritation level.

File: default/AI/DiplomaticCorp.py [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+    # in the meantime, somewhat random
+    if foAI.foAIstate.aggression == fo.aggression.maniacal:
+        return -9
+    elif foAI.foAIstate.aggression == fo.aggression.beginner:
+        return 9
+    log_index = (other_empire_id, fo.empireID())
+    num_peace_requests = len(foAI.foAIstate.diplomatic_logs.get('peace_requests', {}).get(log_index, []))
+    num_war_declarations = len(foAI.foAIstate.diplomatic_logs.get('war_declarations', {}).get(log_index, []))
+    # Too many requests for peace irritate the AI, as do any war declarations
+    irritation = (foAI.foAIstate.aggression * (2.0 + num_peace_requests/10.0 + 2.0 * num_war_declarations) + 0.5)
```
Comment (Cjkjvfnby) [inline default/AI/DiplomaticCorp.py:76]: I don't feel this formula.   IMHO it is better to move such things in separate function and make some examples in docstrings.

```
5 * (2.0 + 0/10 + 2.0 * 0) + 0.5 = 10.5
```

File: default/AI/FreeOrionAI.py [hunk 6]
```diff
@@ -68,8 +69,8 @@ def startNewGame(aggression=fo.aggression.aggressive):  # pylint: disable=invali
     universe = fo.getUniverse()
     if planet_id is not None and planet_id != -1:
         planet = universe.getPlanet(planet_id)
-        new_name = random.choice(_capitols.get(aggression, "").split('\n')).strip() + " " + planet.name
-        print "Capitol City Names are: ", _capitols
+        new_name = " ".join([random.choice(_capitals.get(aggression, []) or [" "]).strip(), planet.name])
```
Comment (Cjkjvfnby) [inline default/AI/FreeOrionAI.py:72]: IMHO this changes is not related to current pr.

File: default/AI/DiplomaticCorp.py [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+        fo.sendChatMessage(proposal_sender_player, acknowledgement)
+        attitude = evaluate_diplomatic_attitude(message.sender)
+        if attitude > 0:
+            reply_text = random.choice(UserStringList("AI_PEACE_PROPOSAL_RESPONSES_YES_" + suffix + "_LIST"))
+            diplo_reply = fo.diplomaticMessage(reply_sender, reply_recipient, fo.diplomaticMessageType.acceptProposal)
+        else:
+            reply_text = random.choice(UserStringList("AI_PEACE_PROPOSAL_RESPONSES_NO_" + suffix + "_LIST"))
+            diplo_reply = None
+        fo.sendChatMessage(proposal_sender_player, reply_text)
+        print "sending chat to player %d of empire %d, message body: '%s'" % \
```
Comment (Cjkjvfnby) [inline default/AI/DiplomaticCorp.py:41]: Other prints inside this function starts from capital letters.
You can move `(` to this string and dont use `\`

File: default/AI/AIstate.py (hunk 1 above)
Comment (Dilvish-fo) [inline default/AI/AIstate.py:1111]: Yes.  At the moment it is just logging incoming war declarations, but I expect to log outgoing war declarations here as well.

Comment (Dilvish-fo) [inline default/AI/AIstate.py:1111]: I don't see how your proposal of how to store the history is really simpler.  It doesn't sound easier to work with.   If you have some use or something that works out better, then talk about those specifics.

File: default/AI/DiplomaticCorp.py (hunk 2 above)
Comment (Dilvish-fo) [inline default/AI/DiplomaticCorp.py:12]: ah, probably not, when I moved the function over from FreeOrionAI I changed the name but forgot to remove the comment.

File: default/AI/DiplomaticCorp.py (hunk 3 above)
Comment (Dilvish-fo) [inline default/AI/DiplomaticCorp.py:20]: I forget if I actually ran into trouble without this, or if it was just preexisting and I figured I'd keep it as 'better safe than sorry'.

File: default/AI/DiplomaticCorp.py (hunk 4 above)
Comment (Dilvish-fo) [inline default/AI/DiplomaticCorp.py:40]: I think it is a reasonable call either way. Once I have the AIs counter declaring war, I initially want it to irritate them slightly to be asked for peace, even if they agree.  Obviously, a more complex treatment wouldn't have to have that.

File: default/AI/DiplomaticCorp.py (hunk 5 above)
Comment (Dilvish-fo) [inline default/AI/DiplomaticCorp.py:76]: pretty much the whole point of this function is to do this formula.   How does it help to move it into another function?  The formula was just worked out to produce a reasonable result, no particular theory behind anything, except to the extent already explained.

File: default/AI/FreeOrionAI.py (hunk 6 above)
Comment (Dilvish-fo) [inline default/AI/FreeOrionAI.py:72]: isn't this part of the first commit, on which the second depends?

File: default/AI/DiplomaticCorp.py (hunk 7 above)
Comment (Dilvish-fo) [inline default/AI/DiplomaticCorp.py:41]: edit: n.m , I saw now the style guide preference for adding parentheses to wrap instead of using \

File: default/AI/FreeOrionAI.py (hunk 6 above)
Comment (Dilvish-fo) [inline default/AI/FreeOrionAI.py:72]: I'll also point out that the entire purpose of having the AIs rename their capitals is to hint/signal their aggression level to the human player(s), and so even if I had been making more substantive changes to the naming it would still fit quite well in a PR about basic diplomacy.

File: default/AI/AIstate.py (hunk 1 above)
Comment (Cjkjvfnby) [inline default/AI/AIstate.py:1111]: Flat `list<tuple>` is more simple then `dict<str, dict<tuple, list>>`

`filter`function  and list comprehension allow to get any information in simple way.

File: default/AI/DiplomaticCorp.py [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+              (proposal_sender_player, reply_recipient, reply_text)
+        if diplo_reply:
+            print "Sending diplomatic message to empire %s of type %s" % (reply_recipient, diplo_reply.type)
+            fo.sendDiplomaticMessage(diplo_reply)
+    elif message.type == fo.diplomaticMessageType.warDeclaration:
+        foAI.foAIstate.log_peace_request(message.sender, message.recipient)
```
Comment (Cjkjvfnby) [inline default/AI/DiplomaticCorp.py]: Make peace not war.

File: default/AI/DiplomaticCorp.py [hunk 9]
```diff
@@ -0,0 +1,78 @@
...
+            diplo_reply = fo.diplomaticMessage(reply_sender, reply_recipient, fo.diplomaticMessageType.acceptProposal)
+        else:
+            reply_text = random.choice(UserStringList("AI_PEACE_PROPOSAL_RESPONSES_NO_" + suffix + "_LIST"))
+            diplo_reply = None
+        fo.sendChatMessage(proposal_sender_player, reply_text)
+        print "sending chat to player %d of empire %d, message body: '%s'" % \
+              (proposal_sender_player, reply_recipient, reply_text)
+        if diplo_reply:
+            print "Sending diplomatic message to empire %s of type %s" % (reply_recipient, diplo_reply.type)
+            fo.sendDiplomaticMessage(diplo_reply)
```
Comment (Cjkjvfnby) [inline default/AI/DiplomaticCorp.py:45]: Should we send Cancel diplomatic message too?

File: default/AI/DiplomaticCorp.py (hunk 8 above)
Comment (Dilvish-fo) [inline default/AI/DiplomaticCorp.py]: Please be less cryptic.  If your suggestion is that the AI sometimes ask for peace, then yes as has been discussed in the forums that is part of the plan.  But that decision is far more complex than what I am trying to accomplish here.

File: default/AI/DiplomaticCorp.py (hunk 9 above)
Comment (Dilvish-fo) [inline default/AI/DiplomaticCorp.py:45]: I'm not even sure if we can cancel a diplomatic offer someone else has sent, and it is probably best we don't.   The clearest result is a (at least temporarily) lost opportunity for the AI which might not get offered again.  Once we get to the point that we have the AI deciding it will bluff for a better offer :smiley: , then we can have it (try to) cancel other people's offers to it.

File: default/AI/DiplomaticCorp.py (hunk 8 above)
Comment (isupeene) [inline default/AI/DiplomaticCorp.py]: I believe he meant that you are using `log_peace_request`, but the message type was `warDeclaration`.  In other words, there is a mismatch between what is logged and what is actually checked for.

**Comment by Cjkjvfnby** [comment]
My version https://github.com/Cjkjvfnby/freeorion/pull/6

**Comment by Morlic-fo** [comment]
I agree with the approach of using subclasses for different personality types (here: aggression levels).

Diplomacy definitely should be an area where personality of the AI matters a lot so I expect the subclasses to be used relatively often to tweak some factors or behaviour. Having all what personality changes in a subclass definitely makes things easier to grasp compared to checking some dozen if-statements scattered around multiple hundred lines of codes.

File: default/AI/DiplomaticCorp.py (hunk 8 above)
Comment (Dilvish-fo) [inline default/AI/DiplomaticCorp.py]: Ah yes, thanks.  Fixed now.

**Comment by Dilvish-fo** [comment]
yes, I agree the class/subclass approach appears very useful here.  I've been pretty busy with some other AI stuff the last few days, but as noted in elsewhere (Cjs PR I think?) I will look into setting the AI aggression level at an earlier point, prior to AIstate being instantiated, so that it can be better handled.

**Comment by Dilvish-fo** [comment]
After looking into it, it seems to me that it's not really necessary to set the AI aggression earlier.  The version committed in f858d61413e51861a8528630f4d3a2e962c023d3 uses the class idea, and instantiates the class once it knows the aggression level.  I left the diplomacy logging essentially the way it was originally proposed, but that's not to necessarily rule out changes along the lines Cj has proposed.
